test: add lib/link.js tests

ruyadorno opened the pull request:
- Added `ELINKGLOBAL` error code for when using `--global`
- Added tests for `lib/link.js`
- ~Do not reify, only load globals when linking a pkg to a local prefix~
- Tweaked `npm link <pkg>` behavior to symlink to available package in global dir if possible, if not reify missing pkg into global dir then reify-symlink it to local dir
- Bumped `@npmcli/arborist@0.0.23`
- Fixes: #1777 

**NOTE:** Tests are not mocking Arborist since I think this is a cmd that might help us out surfacing underlying issues, similar to `npm ls`

ruyadorno committed:
test: add lib/link.js tests

- Added ELINKGLOBAL error code for when using --global
- Added tests for lib/link.js
- Do not reify, only load globals when linking a pkg to a local prefix
- Fixes: #1777

isaacs commented on lib/link.js:
```diff
@@ -1,66 +1,79 @@
   })
 
-  const globals = await globalArb.reify({ add: args })
+  const globals = await globalArb.loadActual()
```
```js
const globals = await globalArb.loadActual({
  filter: (node, kid) => {
    return !node.isRoot || args.some(j => j === kid)
  }
})
```

isaacs commented on lib/link.js:
```diff
@@ -1,66 +1,79 @@
+  const links = [
+    ...globals.children.values()
+  ]
+    .filter(i => args.some(j => j === i.name))
```
```js
const links = [...globals.edgesOut.keys()]
```

isaacs commented on lib/link.js:
```diff
@@ -1,66 +1,79 @@
   })
   await localArb.reify({
-    add: links.map(l => `file:${resolve(globalTop, 'node_modules', l)}`)
+    add: links.map(l => `file:${resolve(globalTop, 'node_modules', l.path)}`)
```
correct.

ruyadorno committed:
fix: reenable npm link from registry

Being able to npm link a package that is not currently available in the
global space should still be a supported feature, this change puts that
functionality back in place but also improves it by avoiding reify any
package that may already be found in the global directory.

ruyadorno committed:
@npmcli/arborist@0.0.23

isaacs commented on lib/link.js:
```diff
@@ -1,66 +1,134 @@
   'link',
   'npm link (in package dir)' +
-  '\nnpm link [<@scope>/]<pkg>[@<version>]'
+  '\nnpm link [<@scope>/]<pkg>[@version]'
```
Pedantic nit, but I think the `<>` is technically correct here, since the `@` is a literal, and `<version>` is a placeholder.

ruyadorno replied:
oops, unintended change 😬 will fix
